ADAM issue #410

bryanjj opened the pull request:
I was interested in contributing to this project and learning scala.
I am attempting to partially address https://github.com/bigdatagenomics/adam/issues/410

I created a new alphabet class which operates on symbols.  I based the functionality on the existing SequenceUtils class which supports the following functionality:
1. basic DNA symbols
2. case-insensitivity
3. reverse complement function which leaves unknown chars unchanged

My main concerns with this pull request are
1. I'm just learning scala so might be doing something dumb
2. I'm not sure which package this class should go in
3. I think performance should be ok because the String map function uses a StringBuilder internally, but not sure about overhead on case class creation or exception handling for unknown characters
4. This issue also mentions "ambiguity codes", but I'm not sure what they are but should be able to add later if they're needed

this is my first github contribution ever so let me know what you think

AmplabJenkins commented:
Can one of the admins verify this patch?

massie commented:
Jenkins, add to whitelist and test this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/ADAM-prb/677/
Test PASSed.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/models/Alphabet.scala:
```diff
@@ -0,0 +1,106 @@
+   * if false, this alphabet will treat upper or lower case chars as the same for its symbols
+   *
+   */
+  val caseSensitive: Boolean
```
+1

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/models/Alphabet.scala:
```diff
@@ -0,0 +1,106 @@
+    if (caseSensitive)
+      symbols.map(x => x.label -> x).toMap
+    else
+      symbols.flatMap(x => Seq(x.label.toLower -> x.toLower, x.label.toUpper -> x.toUpper)).toMap
```
My understanding of the `symbolMap` is that it'll canonicalize an input symbol. In that case, shouldn't the `flatMap`ped Seq be `Seq(x.label.toLower -> x.toUpper, x.label.toUpper -> x.toUpper)`?

bryanjj replied:
Yes I think that makes sense.  The reason I did it this way was to match the existing behavior in the SequenceUtils match statement, where case "a" => "t" and case "A" => "T" but not sure if that is a real requirement

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/models/Alphabet.scala:
```diff
@@ -0,0 +1,106 @@
+   * @return the reversed complement of the given string.
+   */
+  def reverseComplement(s: String): String = {
+    s.map(x => Try(apply(x)).getOrElse(Symbol(x, x)).complement).reverse
```
The `Try` here is to check if char `x` is in your alphabet, correct? If you get an `x` that isn't in the alphabet as input, wouldn't you want to throw an exception? I agree with wrapping in some form of try/catch here, but I would think it to be better to use the try/catch to throw a better exception than the `KeyNotFound` exception you'd get back from the symbol map (e.g., convert the `KeyNotFound` into `throw new IllegalArgumentException("Character %s not found in alphabet.".format(x))`). At least, that's my general thought.

bryanjj replied:
Ok.  so the Try in this case will catch the KeyNotFoundException and replace it with the Symbol(x,x).  This is to maintain the existing behavior in the SequenceUtils class where an unrecognized character is just kept as is.  The reverseComplementExact will throw the KeyNotFound but I can convert that to an IllegalArgument instead

fnothaft commented:
+1, this is good code. Thanks for the contribution @bryanjj! I left a few comments inline, but I think we can merge this soon. As a small nit, can you [squash the commits down](http://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit) into a single commit, and prefix that commit with `[ADAM-410]`?

bryanjj commented:
Thanks for the comments!  I will make the changes and see how it looks

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/688/<h2>Build result: FAILURE</span></h2>GitHub pull request #653 of commit e99cf7ea4b78fdd00f465da00af21644f1354a93 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/653/merge^{commit} # timeout=10 > git branch -a --contains 4e16fab7c8b7a4cded7d6adbd83792df38f9b06b # timeout=10 > git rev-parse remotes/origin/pr/653/merge^{commit} # timeout=10Checking out Revision 4e16fab7c8b7a4cded7d6adbd83792df38f9b06b (origin/pr/653/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 4e16fab7c8b7a4cded7d6adbd83792df38f9b06bFirst time build. Skipping changelog.Triggering ADAM-prb ? 1.0.4,centosTriggering ADAM-prb ? 2.2.0,centosTriggering ADAM-prb ? 2.3.0,centosADAM-prb ? 1.0.4,centos completed with result FAILUREADAM-prb ? 2.2.0,centos completed with result FAILUREADAM-prb ? 2.3.0,centos completed with result FAILURENotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/689/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/690/<h2>Build result: FAILURE</span></h2>GitHub pull request #653 of commit dda0a175f28dc34fa20950146c249286ee6170bd automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/653/merge^{commit} # timeout=10 > git branch -a --contains 793320a9b6763b0ba6a5d63f842f7da0f6ba88cd # timeout=10 > git rev-parse remotes/origin/pr/653/merge^{commit} # timeout=10Checking out Revision 793320a9b6763b0ba6a5d63f842f7da0f6ba88cd (origin/pr/653/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 793320a9b6763b0ba6a5d63f842f7da0f6ba88cdFirst time build. Skipping changelog.Triggering ADAM-prb ? 1.0.4,centosTriggering ADAM-prb ? 2.2.0,centosTriggering ADAM-prb ? 2.3.0,centosADAM-prb ? 1.0.4,centos completed with result SUCCESSADAM-prb ? 2.2.0,centos completed with result SUCCESSADAM-prb ? 2.3.0,centos completed with result FAILURENotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

bryanjj commented:
I addressed the comments.  Please let me know what you think.

After the squash, the build succeeded for hadoop 1.0 and 2.2 but failed for 2.3  I'm not sure if it's related to my change or something random.

fnothaft commented:
Jenkins, retest this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/694/
Test PASSed.

fnothaft commented:
@laserson @tdanford any thoughts on this? I would like to merge by Monday for the 0.17.0 release.

fnothaft commented:
Thanks @bryanjj! I've rebased and manually merged via https://github.com/bigdatagenomics/adam/commit/6ce4ef7c3c23580b1f083642a5c679ff86f59da5.
